compiler/transforms/set_memory_space.py

```diff
@@ -0,0 +1,226 @@
+            ]
+        ):
+            return
+
+        # Function to find/create casting operand if it is necessary
+        def get_cast_op(operand) -> None | memref.MemorySpaceCast:
+            # check if cast is required: must be a memref in wrong memory space
+            if not isinstance(operand, SSAValue):
+                return None
+            if not isinstance(operand.type, memref.MemRefType):
+                return None
+            if (
+                not isinstance(operand.type.memory_space, builtin.IntegerAttr)
+                and operand.type.memory_space.value.data != 1
+            ):
+                return None
+
+            # cast required: find previous cast or create new one
+            cast_op = None
+            for use in operand.uses:
+                if (
+                    isinstance(use.operation, memref.MemorySpaceCast)
+                    and isinstance(use.operation.dest.type, memref.MemRefType)
+                    and use.operation.dest.type.memory_space
+                    == builtin.IntegerAttr(1, builtin.i32)
+                ):
+                    cast_op = use.operation
+                    break
+            # If cast op not found, create and insert new one
+            if cast_op is None:
+                cast_op = memref.MemorySpaceCast.from_type_and_target_space(
+                    operand, operand.type, builtin.IntegerAttr(1, builtin.i32)
+                )
+                rewriter.insert_op_before_matched_op(cast_op)
+
+            return cast_op
+
+        # cast all inputs and outputs to correct memory space
+        new_inputs = [
+            inp if get_cast_op(inp) is None else get_cast_op(inp).dest
+            for inp in op.inputs
+        ]
+        new_outputs = [
+            out if get_cast_op(out) is None else get_cast_op(out).dest
+            for out in op.outputs
+        ]
+
+        # new linalg op with new inputs & outputs
+        linalg_op = linalg.Generic(
+            new_inputs,
+            new_outputs,
+            rewriter.move_region_contents_to_new_regions(op.regions[0]),
+            op.indexing_maps,
+            op.iterator_types,
+            op.doc,
+            op.library_call,
+        )
+
+        # replace op
+        rewriter.replace_matched_op(linalg_op)
+
+
+class RealizeMemorySpaceCasts(RewritePattern):
+    """Realize the inserted memory space casts. In the snitch
+    cluster case, the different clusters can only access their own
+    local TCDM L1 memory, so memory space casts are handled by local
+    allocations and memref copies at the correct time.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.MemorySpaceCast, rewriter: PatternRewriter):
+        # check for invalid memory space cast,
+        # no results => unused, removed by cse
+        if len(op.results) < 1:
+            return
+
+        # remove cast, unused op: will be removed by cse
+        if len(op.results[0].uses) == 0:
```

Comment:
I don't understand this part.
Why is this necessary?
Do you mean that unused parts will be removed? Because then you are running dce (dead code elimination) instead of common subexpression elimination (CSE), no?
I believe DCE is part of `--canonicalize`, but I don't see it used in any of your tests.

Reply:
okay, I thought DCE was part of CSE, but I could be mistaken.
These checks are mainly to make sure the pass does not apply to ops without results because the following lines would then give nasty errors:
```python
op.results[0].type.get_element_type(),
```
I'll update the comments to make this more clear

Reply:
Turns out it is indeed impossible to have a `memref.memory_space_cast` without results, so this test case is not possible. The check is therefore also not necessary, and has been removed